### ISSUE TITLE
Support for chainable APIs aka `This` type

### ISSUE THREAD

**Issue by claudio-silva** [facebook/flow#399]
This code:
```js
/* @flow */

class A {
  x():A {
  	return this;
  }
}

class B extends A {
  y():B {
  	return this;
  }
}

var w = new B();

w.x().y();
```

Gives this error:

```
hello.js:17:1,9: call of method y
Property not found in
  hello.js:3:7,7: A
```
How can I express the return type of `A.x()` to be the same type as the instance's type where `x()` is invoked on?
This is very important for any javascript library that implements a chainable API where elements on the chain may subclass a base class and inherit chainable methods.

**Comment by samwgoldman** [comment]
I don't have a solution here, but this sounds like Objective-C's `instancetype` to me.

@claudio-silva if you were writing in a different statically typed language, how would you write this?

**Comment by claudio-silva** [comment]
Hi @samwgoldman, I don't know Objective-C, but doing a quick search it seems that `instancetype` is something quite similar to this.
As for other examples of this feature on other languages:
The Typescript team is investigating this issue and brainstorming solutions for it, as you can see in [this issue on GitHub](https://github.com/Microsoft/TypeScript/issues/513).
Scala has `this.type`. See [this article](http://scalada.blogspot.pt/2008/02/thistype-for-chaining-method-calls.html).
In PHPDocumentor, you can use the `$this` keyword as a function return type. See [this reference](http://phpdoc.org/docs/latest/guides/types.html#keywords). It is a very useful feature that allows you to get correct autocompletion on IDEs (ex: PHPStorm).
PHPDocumentor also defines `self` and `static` which have other meanings, but that may potentially also be useful for use with Javascript.

**Comment by claudio-silva** [comment]
@samwgoldman I see you have closed this, but no reason was presented. Is this issue solved?

**Comment by samwgoldman** [comment]
@claudio-silva Sorry about that. I closed this because we released support for `this` types in 0.20.0!

**Comment by claudio-silva** [comment]
Great! Thanks!
But I can't find anything about it on the Flow documentation site!
Also, on the site's Blog section (a changelog), you've skipped version 0.20 and went from 0.19 to 0.21, so there is no mention of what's new on version 0.20!!!
Where can I find more info about these new features?

**Comment by samwgoldman** [comment]
Yeah, thanks for calling us out on that. We're actually working on a big docs push, but it's definitely embarrassing that we added a feature and didn't document it. We're putting a lot of effort into the revamp, and are working to ensure that the docs stay updated as we build new features.

We've also not been great at announcing new versions on the blog, but the Changelog.md file in this repo does always get updated and has a sparse breakdown of the changes we've made and features added.
